docs/WritingTests.asciidoc

```diff
@@ -65,7 +65,7 @@ And here is an example in Python:
 -------------------------------------------------------------------
 from testapi import *
 
-def run():
+def run(*_):
```

Comment:
What's the point of this?

Reply:
`def fn(*_)` means ignore all parameters passed to the function.

> We could still document def run(self), as it is possible to call methods on self, e.g. it would be possible to do something like:

That can be done, but I personally wouldn't encourage it, using `self.method_defined_in_py` does a roundtrip to perl and python unnecessarily.

See the following example:

```py
from testapi import *

def run(self):
  diag("output: {}".format(self.helper()))

def helper():
  return "value"
```

it's completely valid and runs (unless `run_args` is not passed of course). But `self.helper` goes to perl to get a handler of the python package to call the helper method FROM perl and return the result in python.

Whereas:

```py
from testapi import *

def run(self):
  diag("output: {}".format(helper()))

def helper():
  return "value"
```

Would stay in python realm for the same operation.

Reply:
They're not affected because they're called from python to perl and not the other way around. I guess that's good news? :sweat_smile:

Reply:
> and the test inherits from it.

Inherits in the perl sense of things, to python... I'm not sure. the "self" argument that gets to the method only has four symbols:

```py
['py_bind_func', 'py_eval', 'py_study_package', 'run'],
```

that `run` is a Perl reference to the Python `run` (inception).

Reply:
I guess I'm getting confused because on the one hand you're saying arguments can't be parsed, and on the other hand *_ would allow any arguments to be parsed.

As a user I would expect it's either no arguments ever or known restrictions e.g. self only or whatever the restrictions are.

Reply:
> I guess I'm getting confused because on the one hand you're saying arguments can't be parsed, and on the other hand *_ would allow any arguments to be parsed.

Yeah, my bad with the terminology (it's very confusing).

Complex objects (Perl blessed References in this case) do not fly properly to python-land. They are accessible but unusable. `run_args` gets to python as: `OpenQA::Test::RunArgs=HASH(0x555970652750)` that can't be:

1. Called (`run_args()`) 
2. Scripted (`run_args['key']`)
3. Iterated (`list(run_args)`)
4. Members called (`run_args.has('abc')`).

Hope it explains better the main reason why dropping the parameters of `run`.

Reply:
So... can we just make it `run(self)` or `run(self, *kwargs)`? My question was aiming for what works rather than what doesn't to be honest

Reply:
I would vote for the latter: `run(self, *_):`. And encourage to keep inside python as much as possible.

Reply:
So what is _* for then?

Reply:
It's being passed if it's defined, the `*_` will drop it in case is passed.

Reply:
It's like `@` in perl signatures, a catchall throwaway basically.

Reply:
Yes... I know what it does. I don't know why we want it here.

Reply:
@kalikiana I'll test what you suggested in os-autoinst.

If it works as it looks like, then we can keep self and be certain that functions in Python have a stable signature.